src/Form/Form.js

```diff
@@ -368,6 +372,23 @@ class Form {
         this.device.scanner.forms.delete(this.form)
     }
 
+    initFormListeners () {
+        // This ensures we fire the handler again if the form is changed
+        this.addListener(this.form, 'input', () => {
+            if (!this.isAutofilling) {
+                this.submitHandlerExecuted = false
+                this.resetShouldPromptToStoreData()
+            }
+        })
+
+        // If it's a form within a shadow tree, attach the submit listener, because it doesn't bubble outside
+        if (this.form instanceof HTMLFormElement && this.form.getRootNode()) {
+            this.addListener(this.form, 'submit', () => {
+                this.submitHandler('in-form submit handler')
+            }, {capture: true})
+        }
```

Comment:
The other listener was simply moved, but this is new. `submit` events don't bubble outside shadow boundaries, so we need to attach the listener to the `form` itself here.